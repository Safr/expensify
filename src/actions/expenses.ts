import { Dispatch, ActionCreator } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { toastr } from 'react-redux-toastr';
import database from '../firebase/firebase';
import { SET_EXPENSES, ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE } from '../actionTypes';
import { asyncActionStart, asyncActionFinish, asyncActionError } from './async';
import StoreState, { ExpensesI } from '../types';

export interface InitExpense {
  type: '@@INIT';
}

export interface AddExpense {
  type: ADD_EXPENSE;
  expense: ExpensesI;
}

export interface EditExpense {
  type: EDIT_EXPENSE;
  id: string;
  updates: any;
}

export interface RemoveExpense {
  type: REMOVE_EXPENSE;
  id: string;
}

export interface SetExpenses {
  type: SET_EXPENSES;
  expenses: ExpensesI[];
}

export type ExpenseAction = InitExpense | AddExpense | EditExpense | RemoveExpense | SetExpenses;

export const addExpense: ActionCreator<AddExpense> = expense => ({
  type: ADD_EXPENSE,
  expense,
});

export const startAddExpense = (expenseData = {}): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
  }: any = expenseData;
    const expense = {
      description, note, amount, createdAt,
    };
    return database.ref(`users/${uid}/expenses`).push(expense).then(ref => {
      toastr.success('Success', 'Expense has been added');
      dispatch(addExpense({
        id: ref.key,
        ...expense,
      }));
    });
  };

export const removeExpense: ActionCreator<RemoveExpense> = (id) => ({
  type: REMOVE_EXPENSE,
  id,
});

export const startRemoveExpense = (id: string): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
    .then(() => {
      toastr.success('Success', 'Expense has been deleted');
      dispatch(removeExpense(id));
    });
  };

export const editExpense: ActionCreator<EditExpense> = (id, updates) => ({
  type: EDIT_EXPENSE,
  id,
  updates,
});

export const startEditExpense = (id: string, updates: ExpensesI): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
    .then(() => {
      toastr.success('Success', 'Expense has been edited');
      dispatch(editExpense(id, updates));
    });
  };

export const setExpenses: ActionCreator<SetExpenses> = expenses => ({
  type: SET_EXPENSES,
  expenses,
});

export const startSetExpenses = (): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    dispatch(asyncActionStart());
    return database.ref(`users/${uid}/expenses`)
    .once('value')
    .then(snapshot => {
      const expenses: ExpensesI[] = [];
      snapshot.forEach((childSnapshot: any) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });
      dispatch(setExpenses(expenses));
      toastr.success('Success', 'Expenses have been loaded');
      dispatch(asyncActionFinish());
    })
    .catch(e => dispatch(asyncActionError()));
  };
