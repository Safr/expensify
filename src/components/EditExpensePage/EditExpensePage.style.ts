import styled from 'styled-components';
import { fontSizeLarge, fontSizeSmall, lSize, mSize, offWhite, sSize } from '../../theme/variables';
import media from '../../theme/media';

export const ButtonWrapper = styled.div`
  display: flex;
  max-width: 620px;
  margin: 0 auto;
  padding-left: 10px;
  padding-right: 10px;
  ${media.tablet`
    margin-bottom: ${lSize};
    padding-left: 20px;
    padding-right: 20px;
    padding-bottom: 0;
    padding-top: 0;
    max-width: 800px;
  `}
`;

const PageHeader = styled.div`
  padding: ${sSize};
  font-size: ${fontSizeSmall};
  background-color: ${offWhite};
  ${media.tablet`
    padding: ${lSize};
    font-size: ${fontSizeLarge};
  `}

  h2 {
    text-align: left;
  }
`;

export default PageHeader;