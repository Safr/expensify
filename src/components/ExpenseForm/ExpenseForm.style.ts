import styled, { css } from 'styled-components';
import {
  blue,
  borderColor,
  darkRed, fontSizeLarge, fontSizeSmall, lSize, mSize, sSize, white } from '../../theme/variables';
import media from '../../theme/media';

export const InputGroup = styled.form`
  display: flex;
  flex-direction: column;
  max-width: 620px;
  margin: 0 auto;
  padding-left: 10px;
  padding-right: 10px;
  padding-top: 15px;
  margin-bottom: ${mSize};
  ${media.tablet`
    margin-bottom: ${lSize};
    padding-left: 20px;
    padding-right: 20px;
    padding-bottom: 0;
    max-width: 800px;
  `}
`;

export const TextInput = styled.input`
  height: 50px;
  margin-bottom: ${sSize};
  padding: ${sSize};
  font-size: ${fontSizeLarge};
  font-weight: 300;
  border: 1px solid #cacccd;
`;

export const TextArea = styled.textarea`
  height: 100px;
  margin-bottom: ${sSize};
  margin-top: ${sSize};
  padding: ${sSize};
  font-size: ${fontSizeLarge};
  font-weight: 300;
  border: 1px solid ${borderColor};
`;

export const Paragraph = styled.p`
  margin: 0 0 ${mSize} 0;
  font-style: italic;
`;

export const Button = styled.button`
  display: inline-block;
  padding: ${sSize};
  margin-top: ${mSize};
  margin-right: auto;
  font-size: ${fontSizeSmall};
  font-weight: 300;
  line-height: 1;
  color: ${white};
  background: ${blue};
  ${({ red }) => red && css`
    background: ${darkRed};
  `}
  border-radius: 3px;
  border: none;
  ${media.tablet`
    font-size: ${fontSizeLarge};
  `}
` as any;

