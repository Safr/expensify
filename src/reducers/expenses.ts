import { SET_EXPENSES, ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE } from '../actionTypes';
import { ExpenseAction } from '../actions';
import { ExpensesI } from '../types';

const expensesReducer = (state = [], action: ExpenseAction): ExpensesI | ExpensesI[] => {
  switch (action.type) {
    case SET_EXPENSES:
      return [...action.expenses];
    case ADD_EXPENSE:
      return [
        ...state,
        action.expense,
      ];
    case REMOVE_EXPENSE:
      return state.filter(({ id }: {id: string}) => id !== action.id);
    case EDIT_EXPENSE:
      return state.map(expense => (
        expense.id === action.id ? { ...expense, ...action.updates } : expense));
    default:
      return state;
  }
};

export default expensesReducer;

