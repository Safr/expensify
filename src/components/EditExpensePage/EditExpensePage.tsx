import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import ExpenseForm from '../ExpenseForm/ExpenseForm';
import { Button } from '../ExpenseForm/ExpenseForm.style';
import { startEditExpense, startRemoveExpense } from '../../actions';
import StoreState, { ExpensesI, RouteComponentProps } from '../../types';
import PageHeader, { ButtonWrapper } from './EditExpensePage.style';

interface Props extends RouteComponentProps<any> {
  expense: ExpensesI;
  onStartEditExpense: (id: string, expense: ExpensesI) => void;
  onStartRemoveExpense: (id: string) => void;
}

export class EditExpensePage extends Component<Props, {}> {
  handleSubmit = (expense: any) => {
    this.props.onStartEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  handleRemove = () => {
    const id = this.props.expense.id;
    this.props.onStartRemoveExpense(id);
    this.props.history.push('/');
  }
  render() {
    return (
      <>
        <PageHeader>
          <h1>Edit Expense</h1>
        </PageHeader>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.handleSubmit}
        />
        <ButtonWrapper>
          <Button red onClick={this.handleRemove}>
            Remove Expense
          </Button>
        </ButtonWrapper>
      </>
    );
  }
}

const mapStateToProps = (state: StoreState, ownProps: Props) => ({
  expense: state.expenses.find(expense => expense.id === ownProps.match.params.id),
});

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  onStartEditExpense: (id: string, expense: ExpensesI) => dispatch(startEditExpense(id, expense)),
  onStartRemoveExpense: (id: string) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
