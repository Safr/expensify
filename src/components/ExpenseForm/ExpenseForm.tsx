import React, { Component, ChangeEvent, FormEvent } from 'react';
import moment, { Moment } from 'moment';
import { SingleDatePicker } from 'react-dates';
import { ExpensesI } from '../../types';
import { Button, InputGroup, TextArea, TextInput, Paragraph } from './ExpenseForm.style';

interface Props {
  expense?: ExpensesI;
  onSubmit: (expense: object) => void;
}

interface State {
  description: string;
  note: string;
  amount: number | string;
  createdAt: Moment;
  calendarFocused: boolean;
  error: string;
}

class ExpenseForm extends Component<Props, State> {
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: '',
  };

  handleDescriptionChange = (e: ChangeEvent<HTMLInputElement>) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  handleNoteChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  handleDateChange = (createdAt: Moment) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  handleFocusChange = ({ focused }: {focused: boolean}) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please, provide description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  }

  render() {
    return (
      <InputGroup onSubmit={this.handleSubmit}>
        {this.state.error && <Paragraph>{this.state.error}</Paragraph>}
        <TextInput
          type="text"
          placeholder="Description"
          value={this.state.description}
          onChange={this.handleDescriptionChange}
        />
        <TextInput
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.handleAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          focused={this.state.calendarFocused}
          numberOfMonths={1}
          // tslint:disable-next-line:jsx-no-lambda
          isOutsideRange={() => false}
          onDateChange={this.handleDateChange}
          onFocusChange={this.handleFocusChange}
        />
        <TextArea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.handleNoteChange}
        />
        <Button>Save Expense</Button>
      </InputGroup>
    );
  }
}

export default ExpenseForm;
