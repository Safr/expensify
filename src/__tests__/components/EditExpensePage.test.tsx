import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage/EditExpensePage';
import expenses from '../../fixtures/expenses';
import { Button } from '../../components/ExpenseForm/ExpenseForm.style';

describe('EditExpensePage component', () => {
  let onStartEditExpense;
  let onStartRemoveExpense;
  let history;
  let wrapper;
  beforeEach(() => {
    onStartEditExpense = jest.fn();
    onStartRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    // tslint:disable-next-line:jsx-wrap-multiline
    wrapper = shallow(<EditExpensePage
      expense={expenses[0]}
      onStartEditExpense={onStartEditExpense}
      onStartRemoveExpense={onStartRemoveExpense}
      history={history}
    />);
  });
  test('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('should handle onStartEditExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onStartEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  });
  test('should handle onStartRemoveExpense', () => {
    const id = expenses[0].id;
    wrapper.find(Button).simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onStartRemoveExpense).toHaveBeenLastCalledWith(id);
  });
});
