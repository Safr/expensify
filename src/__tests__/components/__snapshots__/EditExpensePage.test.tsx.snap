// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage component should render EditExpensePage correctly 1`] = `
<React.Fragment>
  <styled.div>
    <h1>
      Edit Expense
    </h1>
  </styled.div>
  <ExpenseForm
    expense={
      Object {
        "amount": 77,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <styled.div>
    <styled.button
      onClick={[Function]}
      red={true}
    >
      Remove Expense
    </styled.button>
  </styled.div>
</React.Fragment>
`;
